src/EventSubscriber/GenericEventSubscriber.php

```diff
@@ -78,21 +91,28 @@ public function onRulesEvent(Event $event, $event_name) {
     // @todo Only load active reaction rules here.
     $configs = $storage->loadByProperties(['event' => $event_name]);
 
-    // Loop over all rules and execute them.
-    foreach ($configs as $rules_config) {
-      $reaction_rule = $rules_config->getExpression();
-
-      $context_names = array_keys($reaction_rule->getContextDefinitions());
-      foreach ($context_names as $context_name) {
-        // If this is a GenericEvent get the context for the rule from the event
-        // arguments.
-        if ($event instanceof GenericEvent) {
-          $reaction_rule->setContextValue($context_name, $event->getArgument($context_name));
-        }
+    // Set up an execution state with the event context.
+    $event_definition = $this->eventManager->getDefinition($event_name);
+    $state = RulesState::create();
+    foreach ($event_definition['context'] as $context_name => $context_definition) {
+      // If this is a GenericEvent get the context for the rule from the event
+      // arguments.
+      if ($event instanceof GenericEvent) {
+        $state->addVariable(
+          $context_name,
+          $context_definition,
+          $event->getArgument($context_name)
+        );
       }
+    }
 
-      $reaction_rule->execute();
+    // Loop over all rules and execute them.
+    foreach ($configs as $config) {
+      /** @var ReactionRule $config */
+      $config->getExpression()
+        ->executeWithState($state);
```

Comment:
Tried fixing it, but turns out we miss that functinoality still in d8. Thus, I created an issue instead: https://www.drupal.org/node/2631948 We can work on this as part of the loop or component execution derivative support tasks.